File: src/Romarr.Core/Games/Game.cs
using System;
using System.Collections.Generic;
using Romarr.Common.Extensions;
using Romarr.Core.Datastore;
using Romarr.Core.Languages;
using Romarr.Core.Profiles.Qualities;

namespace Romarr.Core.Games
{
    public class Game : ModelBase
    {
        public Game()
        {
            Images = new List<MediaCover.MediaCover>();
            Genres = new List<string>();
            Actors = new List<Actor>();
            Platforms = new List<Platform>();
            Tags = new HashSet<int>();
            OriginalLanguage = Language.English;
            MalIds = new HashSet<int>();
            AniListIds = new HashSet<int>();
            PreferredRegions = new List<string>();
            PreferredLanguageIds = new List<int>();
            PreferredReleaseTypes = new List<string>();
            PreferredModifications = new List<string>();
        }

        public int IgdbId { get; set; }
        public int MobyGamesId { get; set; }
        public int RawgId { get; set; }
        public string ImdbId { get; set; }
        public int TmdbId { get; set; }
        public HashSet<int> MalIds { get; set; }
        public HashSet<int> AniListIds { get; set; }
        public string Title { get; set; }
        public string CleanTitle { get; set; }
        public string SortTitle { get; set; }
        public GameStatusType Status { get; set; }
        public string Overview { get; set; }
        public string AirTime { get; set; }
        public bool Monitored { get; set; }
        public NewItemMonitorTypes MonitorNewItems { get; set; }
        public int QualityProfileId { get; set; }
        public bool PlatformFolder { get; set; }
        public DateTime? LastInfoSync { get; set; }
        public int Runtime { get; set; }
        public List<MediaCover.MediaCover> Images { get; set; }
        public GameTypes GameType { get; set; }
        public string Network { get; set; }
        public bool UseSceneNumbering { get; set; }
        public string TitleSlug { get; set; }
        public string Path { get; set; }
        public int Year { get; set; }
        public Ratings Ratings { get; set; }
        public List<string> Genres { get; set; }
        public List<Actor> Actors { get; set; }
        public string Certification { get; set; }
        public string RootFolderPath { get; set; }
        public DateTime Added { get; set; }
        public DateTime? FirstAired { get; set; }
        public DateTime? LastAired { get; set; }
        public LazyLoaded<QualityProfile> QualityProfile { get; set; }
        public Language OriginalLanguage { get; set; }
        public string OriginalCountry { get; set; }
        public List<Platform> Platforms { get; set; }
        public HashSet<int> Tags { get; set; }
        public AddGameOptions AddOptions { get; set; }
        public int? GameSystemId { get; set; }
        public List<string> PreferredRegions { get; set; }
        public List<int> PreferredLanguageIds { get; set; }
        public List<string> PreferredReleaseTypes { get; set; }
        public List<string> PreferredModifications { get; set; }

        public override string ToString()
        {
            return string.Format("[{0}][{1}]", IgdbId, Title.NullSafe());
        }

        public void ApplyChanges(Game otherGame)
        {
            IgdbId = otherGame.IgdbId;

            Platforms = otherGame.Platforms;
            Path = otherGame.Path;
            QualityProfileId = otherGame.QualityProfileId;

            PlatformFolder = otherGame.PlatformFolder;
            Monitored = otherGame.Monitored;
            MonitorNewItems = otherGame.MonitorNewItems;

            GameType = otherGame.GameType;
            RootFolderPath = otherGame.RootFolderPath;
            Tags = otherGame.Tags;
            AddOptions = otherGame.AddOptions;
            GameSystemId = otherGame.GameSystemId;
            PreferredRegions = otherGame.PreferredRegions;
            PreferredLanguageIds = otherGame.PreferredLanguageIds;
            PreferredReleaseTypes = otherGame.PreferredReleaseTypes;
            PreferredModifications = otherGame.PreferredModifications;
        }
    }
}

File: frontend/src/Game/Game.ts
import ModelBase from 'App/ModelBase';
import Language from 'Language/Language';

export type GameType = 'standard';
export type GameMonitor =
  | 'all'
  | 'future'
  | 'missing'
  | 'existing'
  | 'firstPlatform'
  | 'lastPlatform'
  | 'baseGame'
  | 'allDlcs'
  | 'latestUpdate'
  | 'allAdditional'
  | 'none';

export type GameStatus = 'released' | 'continuing' | 'ended' | 'upcoming' | 'deleted';

export type MonitorNewItems = 'all' | 'none';

export type CoverType = 'poster' | 'banner' | 'fanart' | 'platform';

export interface Image {
  coverType: CoverType;
  url: string;
  remoteUrl: string;
}

export interface Statistics {
  platformCount: number;
  fileCount: number;
  downloadedFileCount: number;
  percentOfRoms: number;
  previousAiring?: Date;
  releaseGroups: string[];
  sizeOnDisk: number;
  totalFileCount: number;
  monitoredFileCount: number;
  lastAired?: string;
}

export interface Platform {
  monitored: boolean;
  platformNumber: number;
  title?: string;
  statistics: Statistics;
}

export interface Ratings {
  votes: number;
  value: number;
}

export interface AlternateTitle {
  platformNumber: number;
  scenePlatformNumber?: number;
  title: string;
  sceneOrigin: 'unknown' | 'unknown:igdb' | 'mixed' | 'igdb';
  comment?: string;
}

export interface GameAddOptions {
  monitor: GameMonitor;
  searchForMissingRoms: boolean;
  searchForCutoffUnmetRoms?: boolean;
}

interface Game extends ModelBase {
  added: string;
  alternateTitles: AlternateTitle[];
  certification: string;
  cleanTitle: string;
  ended: boolean;
  firstAired: string;
  genres: string[];
  images: Image[];
  imdbId?: string;
  monitored: boolean;
  monitorNewItems: MonitorNewItems;
  network: string;
  originalCountry: string;
  originalLanguage: Language;
  overview: string;
  path: string;
  previousAiring?: string;
  nextAiring?: string;
  qualityProfileId: number;
  ratings: Ratings;
  rootFolderPath: string;
  runtime: number;
  platformFolder: boolean;
  platforms: Platform[];
  gameType: GameType;
  sortTitle: string;
  statistics?: Statistics;
  status: GameStatus;
  tags: number[];
  title: string;
  titleSlug: string;
  igdbId: number;
  rawgId: number;
  mobyGamesId: number;
  tmdbId: number;
  useSceneNumbering: boolean;
  year: number;
  addOptions: GameAddOptions;
  gameSystemId?: number;
  preferredRegions: string[];
  preferredLanguageIds: number[];
  preferredReleaseTypes: string[];
  preferredModifications: string[];
}

export default Game;
